src/CodeGen_PyTorch.cpp

```diff
@@ -0,0 +1,395 @@
+            if (type.is_uint()) {
+                oss << 'u';
+            }
+            oss << "int" << type.bits();
+            if (type.is_vector()) {
+                oss << "x" << type.lanes();
+            }
+            oss << "_t";
+            break;
+        default:
+            user_error << "Can't represent an integer with this many bits in C: " << type << "\n";
+        }
+    }
+    if (include_space && needs_space)
+        oss << " ";
+    return oss.str();
+}
+
+string type_to_pytorch_tensor(Type type, bool isCuda) {
+    return "at::Tensor";
+}
+
+} // ns anon
+
+
+CodeGen_PyTorch::CodeGen_PyTorch(ostream &s, Target t, OutputKind output_kind,
+    std::string cpp_header) :
+    IRPrinter(s), target(t), output_kind(output_kind), cpp_header(cpp_header)
+{
+  // TODO(mgharbi): header guard and header / implementation split
+  stream << "#include <torch/extension.h>\n";
+  // TODO(mgharbi): find a shallower integration with torch cuda, handle no GPU case
+  stream << "#include <HalideBuffer.h>\n\n";
+
+  // Conditionally add CUDA features to the pytorch helper
+  // if(target.has_feature(Target::CUDA)) {
+  //   stream << "#define HL_PT_CUDA\n";
+  // }
+  stream << "#include <HalidePytorchHelpers.h>\n";
+  if(target.has_feature(Target::CUDA)) {
+    stream << "#include <ATen/cuda/CUDAContext.h>\n";  
+    stream << "#include <HalidePytorchCudaHelpers.h>\n";
+    // stream << "#undef HL_PT_CUDA\n";
+  }
+
+  std::vector<std::string> header_path = split_string(cpp_header, "/");
+  std::string header = header_path.back();
+  
+
+  stream << "\n#include \"" << header << "\"\n\n";
+  stream << "using Halide::Runtime::Buffer;\n\n";
+}
+
+CodeGen_PyTorch::~CodeGen_PyTorch() {
+}
+
+void CodeGen_PyTorch::compile(const Module &input) {
+    for (const auto &f : input.functions()) {
+      if (f.name.find("old_buffer_t") != std::string::npos) {
+        debug(1) << "ignoring " << f.name;
+        continue;
+      }
+      if(target.has_feature(Target::CUDA)) {
+        compile(f, true);
+      } else {
+        compile(f, false);
+      }
+    }
+}
+
+void CodeGen_PyTorch::compile(const LoweredFunc &f, bool isCuda) {
+  // Don't put non-external function declarations in headers.
+  std::vector<std::string> namespaces;
+  std::string simple_name = extract_namespaces(f.name, namespaces);
+
+  if (!namespaces.empty()) {
+    for (const auto &ns : namespaces) {
+      stream << "namespace " << ns << " {\n";
+    }
+    stream << "\n";
+  }
+  const std::vector<LoweredArgument> &args = f.args;
+  std::vector<LoweredArgument> buffer_args;
+
+  stream << "int " << simple_name << "_th_(";
+  for (size_t i = 0; i < args.size(); i++) {
+    if (args[i].name == "__user_context") {
+      continue;
+    } else if (args[i].is_buffer()) {
+      buffer_args.push_back(args[i]);
+      // if(args[i].is_input()) { // TODO
+      //   stream << "const ";
+      // }
+      stream 
+        << type_to_pytorch_tensor(args[i].type, isCuda)
```

Comment:
style used in Halide would suggest `is_cuda` rather than `isCuda`

Reply:
updated that in afd9747